src/util/sentinel.hpp

```diff
@@ -163,6 +161,16 @@ class sentinel_iterator {
     bool operator>(const sentinel_iterator& x) const {
         return !(x<=*this);
     }
+
+    // access to underlying iterator/sentinel
+
+    bool is_sentinel() const { return e_.index()!=0; }
+    bool is_iterator() const { return e_.index()==0; }
```

Comment:
This form reflects the state of the underlying variant type, but I'm not strongly committed to this form instead of what you suggest.